### DISCUSSION TITLE
ADC data abnormal in some channel for LPC55s69

### DISCUSSION THREAD

**Discussion by MarsMSWu** [zephyrproject-rtos/zephyr#68088] [Q&A] [answered]
Hi, all
  I have some question for the ADC channel.
  Currently, I use ADC define in dts file as below :

    zephyr,user {
        io-channels = <&adc0 0 &adc0 1 &adc0 2 &adc0 3 &adc0 4 &adc0 8 &adc0 11 &adc0 12>;
    };

	channel@0 {
		reg = <0>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_EXTERNAL0";
		zephyr,vref-mv = <3300>;
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <16>;
		zephyr,input-positive = <MCUX_LPADC_CH0A>;
	};

    channel@1 {
		reg = <1>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_EXTERNAL0";
		zephyr,vref-mv = <3300>;
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <16>;
		zephyr,input-positive = <MCUX_LPADC_CH1A>;
	};

	channel@2 {
		reg = <2>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_EXTERNAL0";
		zephyr,vref-mv = <3300>;
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <16>;
		zephyr,input-positive = <MCUX_LPADC_CH2A>;
	};

    channel@3 {
		reg = <3>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_EXTERNAL0";
		zephyr,vref-mv = <3300>;
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <16>;
		zephyr,input-positive = <MCUX_LPADC_CH3A>;
	};

    channel@4 {
        reg = <4>;
        zephyr,gain = "ADC_GAIN_1";
        zephyr,reference = "ADC_REF_EXTERNAL0";
        zephyr,vref-mv = <3300>;
        zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
        zephyr,resolution = <16>;
        zephyr,input-positive = <MCUX_LPADC_CH4A>;
    };

	channel@8 {
		reg = <8>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_EXTERNAL0";
		zephyr,vref-mv = <3300>;
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <16>;
		zephyr,input-positive = <MCUX_LPADC_CH0B>;
    };

	channel@11 {
		reg = <11>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_EXTERNAL0";
		zephyr,vref-mv = <3300>;
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <16>;
		zephyr,input-positive = <MCUX_LPADC_CH3B>;
	};

	channel@12 {
		reg = <12>;
		zephyr,gain = "ADC_GAIN_1";
		zephyr,reference = "ADC_REF_EXTERNAL0";
		zephyr,vref-mv = <3300>;
		zephyr,acquisition-time = <ADC_ACQ_TIME_DEFAULT>;
		zephyr,resolution = <16>;
		zephyr,input-positive = <MCUX_LPADC_CH4B>;
	};

	pinmux_adc0_default: pinmux_adc0_default {
		group0 {
			pinmux = <ADC0_CH0_PIO0_23>,
				<ADC0_CH1_PIO0_10>,
				<ADC0_CH2_PIO0_15>,
				<ADC0_CH3_PIO0_31>,
				<ADC0_CH4_PIO1_8>,
				<ADC0_CH8_PIO0_16>,
				<ADC0_CH11_PIO1_0>,
				<ADC0_CH12_PIO1_9>;
			slew-rate = "standard";
			nxp,analog-mode;
		};
	};

  In my real test, only ADC0_12, ADC0_8 and ADC0_1 can get positive result.
  These three channel use default external pull-high 3.3V and can get digital value 65xxx in my console log.
  If I force GND these three channel and can get 0~10 value.
  So I think at least these three channel configuration should be correct.
  I add other ADC channels and set related channel and side, other setting are the same but result is not reach my expect.
  I don't know is need any extra configure or maybe limitation...

Thanks.

**Comment by DerekSnell** [comment]
Hi @MarsMSWu ,
Your devicetree looks fine to me.  In fact, I used your devicetree as the attached 
[lpcxpresso55s69_cpu0.overlay](https://github.com/zephyrproject-rtos/zephyr/files/14090638/lpcxpresso55s69_cpu0.zip), and I tested on the LPCXpresso55S69 board.  Each ADC channel measured as expected for me.  To test, I used the overlay with the [ADC driver sample](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/drivers/adc).

Can you try that sample and see how it compares for you?  Let us know what you find.

The console output:
```
*** Booting Zephyr OS build zephyr-v3.5.0-4933-g70809cc429cc ***                
ADC reading[0]:                                                                 
- adc@A0000, channel 0: 24837 = 1250 mV                                         
- adc@A0000, channel 1: 57933 = 2917 mV                                         
- adc@A0000, channel 2: 37633 = 1894 mV                                         
- adc@A0000, channel 3: 53145 = 2676 mV                                         
- adc@A0000, channel 4: 38342 = 1930 mV                                         
- adc@A0000, channel 8: 35203 = 1772 mV                                         
- adc@A0000, channel 11: 55485 = 2793 mV                                        
- adc@A0000, channel 12: 56591 = 2849 mV   
```

**Reply by MarsMSWu** [reply, marked as answer]
Hi @DerekSnell ,

It 's very thanks for your feedback.
I think maybe is wrong data type then also get a incorrect result.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/db342205-44b3-408c-8845-6a199350432d)
We use a integer type to get the data and put to unsigned integer type and conversion to mV.
I don't know why buf data is negative...then if assign to unsigned integer type will get a huge and incorrect result.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/9bb487ff-9764-45b1-8559-f6b177d05c7a)
If we use unsigned integer data type for buf then can get correcr result.
![image](https://github.com/zephyrproject-rtos/zephyr/assets/151608605/485d4c47-5a17-4eba-9ca3-9a87e58a341d)

Because that channel is using external pull-hi 3.3V so print the value in console is 65xxx, it meet our expected in early study.
The value 65xxx show on console should be mV data type which conversion by API function "**_adc_raw_to_millivolts_dt_**".

Thanks.
